File: EC/pony_gp.js
/**
 * Created by erikhemberg on 7/16/14.
 */

// TODO refactor for oop

var DEFAULT_FITNESS = 1000;

// From http://indiegamr.com/generate-repeatable-random-numbers-in-js/
// the initial seed
Math.seed = 711;
Math.seededRandom = function (max, min) {
    Math.seed = (Math.seed * 9301 + 49297) % 233280;
    var rnd = Math.seed / 233280;
    return min + rnd * (max - min);
};

function copy_tree(tree) {
    var out = [];
    for (var i = 0; i < tree.length; i++) {
        if (typeof tree[i] === 'string') {
            out = tree;
        } else {
            out[i] = copy_tree(tree[i]);
        }
    }
    return out.slice(0);
}

function tree_to_str(tree) {
    return "(" + node_to_str(tree, "") + ")"
}

function node_to_str(tree, str) {
    if (typeof tree === 'string') {
        str += tree;
    } else {
        str += tree[0];
        for (var i = 1; i < tree.length; i++) {
            str += ",(";
            str = node_to_str(tree[i], str);
            str += ")";
        }
    }
    return str;
}

function get_symbols() {
    var arity = {
        1: 0,
        0: 0,
        "+": 2,
        "*": 2
    };

    var terminals = [];
    var functions = [];

    for (var key in arity) {
        if (arity.hasOwnProperty(key)) {
            if (arity[key] == 0) {
                terminals.push(key);
            } else {
                functions.push(key);
            }
        }
    }

    return {arity: arity, terminals: terminals, functions: functions};
}

var symbols = get_symbols();

function get_random() {
    return Math.seededRandom(0, 1);
}
function get_random_int(min, max) {
    return Math.floor(Math.seededRandom(min, max));
}

function get_random_boolean() {
    return Math.seededRandom(0, 1) < 0.5;
}

function get_random_symbol(depth, max_depth, symbols, full) {
    var symbol;
    if (depth >= (max_depth - 1)) {
        symbol = symbols["terminals"][get_random_int(0, symbols["terminals"].length)];
    } else {
        var terminal = get_random_boolean();
        if (!full && terminal) {
            symbol = symbols["terminals"][get_random_int(0, symbols["terminals"].length)];
        } else {
            symbol = symbols["functions"][get_random_int(0, symbols["functions"].length)];
        }
    }
    return symbol;
}

function sum(array) {
    return array.reduce(function (previous_value, current_value) {
        return previous_value + current_value;
    });
}

function sort_individuals(individual_0, individual_1) {
    if (individual_0["fitness"] < individual_1["fitness"]) {
        return 1;
    }
    if (individual_0["fitness"] > individual_1["fitness"]) {
        return -1;
    }
    return 0;
}

function evaluate(tree) {
    // Get the symbol of the node
    var symbol;
    if (typeof tree === 'string') {
        symbol = tree;
    } else {
        symbol = tree[0];
    }
    var value;
    if (symbol == "+") {
        value = evaluate(tree[1]) + evaluate(tree[2]);
    } else if (symbol == "*") {
        value = evaluate(tree[1]) * evaluate(tree[2]);
    } else if (symbol == "0") {
        value = Number(symbol);
    } else if (symbol == "1") {
        value = Number(symbol);
    } else {
        throw "Unknown symbol:" + symbol;
    }
    return value;
}

function evaluate_individual(individual) {
    return evaluate(individual);
}

function evaluate_fitness(population) {
    // Evaluate fitness
    for (var i = 0; i < population.length; i++) {
        population[i]["fitness"] = evaluate_individual(population[i]["genome"]);
    }
}

function get_ave_and_std(values) {
    var ave = sum(values) / values.length;
    var std = 0;
    for (var val in values) {
        std = std + Math.pow((val - ave), 2);
    }
    std = Math.sqrt(std / values.length);
    return [ave, std];
}

function get_max_tree_depth(node, depth, max_depth) {
    if (max_depth < depth) {
        max_depth = depth;
    }
    for (var i = 1; i < node.length; i++) {
        depth = depth + 1;
        max_depth = get_max_tree_depth(node[i], depth, max_depth);
        depth = depth - 1;
    }
    return max_depth;
}

function max(list) {
    return list.reduce(function(previous, current) {
        return previous > current ? previous : current;
    })
}

function min(list) {
    return list.reduce(function(previous, current) {
        return previous < current ? previous : current;
    })
}

function print_stats(generation, population) {
    population.sort(sort_individuals);
    var fitness_values = [];
    var sizes = [];
    var depths = [];
    for (var i = 0; i < population.length; i++) {
        fitness_values.push(population[i]["fitness"]);
        sizes.push(get_number_of_nodes(population[i]["genome"], 0));
        depths.push(get_max_tree_depth(population[i]["genome"], 0, 0));
    }
    var ave_and_std = get_ave_and_std(fitness_values);
    var ave_and_std_size = get_ave_and_std(sizes);
    var ave_and_std_depth = get_ave_and_std(depths);
    console.log("Gen:" + generation + " fit_ave:" + ave_and_std[0] + "+-" + ave_and_std[1] +
        " size_ave:" + ave_and_std_size[0] + "+-" + ave_and_std_size[1] +
        " depth_ave:" + ave_and_std_depth[0] + "+-" + ave_and_std_depth[1] +
        " " + population[0]["fitness"] + " " + tree_to_str(population[0]["genome"]));
    console.log("min_fit:" + min(fitness_values) + " max_fit:" + max(fitness_values) +
        " min_size:" + min(sizes) + " max_size:" + max(sizes) +
        " min_depth:" + min(depths) + " max_depth:" + max(depths))
}

function tournament_selection(tournament_size, population) {
    var new_population = [];
    while (new_population.length < population.length) {
        var competitors = [];
        // Randomly select competitors for the tournament
        for (var i = 0; i < tournament_size; i++) {
            var idx = get_random_int(0, population.length);
            competitors.push(population[idx]);
        }
        // Sort the competitors by fitness
        competitors.sort(sort_individuals);
        // Push the best competitor to the new population
        var winner = competitors[0]["genome"];
        winner = copy_tree(winner);
        new_population.push({genome: winner, fitness: DEFAULT_FITNESS});
    }
    return new_population;
}

function get_number_of_nodes(root, cnt) {
    cnt = cnt + 1;
    if (typeof root !== 'string') {
        for (var i = 1; i < root.length; i++) {
            cnt = get_number_of_nodes(root[i], cnt);
        }
    }
    return cnt;
}

function get_depth_from_index(node, tree_info, node_idx, depth) {

    if (node_idx == tree_info["idx"]) {
        tree_info["idx_depth"] = depth;
    }
    tree_info["idx"] = tree_info["idx"] + 1;
    for (var i = 1; i < node.length; i++) {
        depth = depth + 1;
        tree_info = get_depth_from_index(node[i], tree_info, node_idx, depth);
        depth = depth - 1;
    }
    return tree_info;
}

function find_and_replace_subtree(root, subtree, node_idx, idx) {
    idx = idx + 1;
    if (node_idx == idx) {
        replace_subtree(subtree, root);
    } else {
        for (var i = 1; i < root.length; i++) {
            idx = find_and_replace_subtree(root[i], subtree, node_idx, idx);
        }
    }
    return idx;
}

function mutation(mutation_probability, new_population, max_size) {
    for (var i = 0; i < new_population.length; i++) {
        // Mutate individuals
        if (get_random() < mutation_probability) {
            var end_node_idx = get_number_of_nodes(new_population[i]["genome"], 0) - 1;
            var node_idx = get_random_int(0, end_node_idx);
            var node_info = get_depth_from_index(new_population[i]["genome"], {idx_depth: 0, idx: 0}, node_idx, 0);
            var max_subtree_depth = max_size - node_info['idx_depth'];
            //console.log('mut', max_subtree_depth, max_size, node_info['idx_depth'], new_population[i]["genome"]);
            var new_subtree = [get_random_symbol(max_subtree_depth, max_size, symbols)];
            if (contains(symbols['functions'], new_subtree[0])) {
                var full = false;//get_random_boolean();
                grow(new_subtree, node_info['idx_depth'], max_size, full, symbols);
            }
            find_and_replace_subtree(new_population[i]["genome"], new_subtree, node_idx, -1);
            var new_depth = get_max_tree_depth(new_population[i]["genome"], 0, 0);
            //console.log('new depth:', new_depth, new_population[i]["genome"]);
        }
    }
}

function contains(array, obj) {
    for (var i = 0; i < array.length; i++) {
        if (array[i] === obj) {
            return true;
        }
    }
    return false;
}

function append_symbol(node, symbol, symbols) {
    var new_node;
    if (contains(symbols['terminals'], symbol)) {
        new_node = symbol;
    } else {
        new_node = [symbol];
    }
    node.push(new_node);
    return new_node
}

function grow(tree, depth, max_depth, full, symbols) {
    var symbol;
    if (typeof tree !== 'string') {
        symbol = tree[0];
    } else {
        symbol = tree;
    }
    for (var i = 0; i < symbols["arity"][symbol]; i++) {
        var new_symbol = get_random_symbol(depth, max_depth, symbols, full);
        var new_node = append_symbol(tree, new_symbol, symbols);
        var new_depth = depth + 1;
        if (contains(symbols['functions'], new_symbol)) {
            grow(new_node, new_depth, max_depth, full, symbols);
        }
    }
}

function initialize_population(population_size, max_size) {
    var population = [];
    for (var i = 0; i < population_size; i++) {
        var full = get_random_boolean();
        var max_depth = (i % max_size) + 1;
        var symbol = get_random_symbol(1, max_depth, symbols, full);
        var tree = [symbol];
        if (max_depth > 0 && contains(symbols["functions"], symbol)) {
            grow(tree, 1, max_depth, full, symbols);
        }
        population.push({genome: tree, fitness: DEFAULT_FITNESS});
        console.log(i, tree_to_str(population[i]["genome"]));
        console.log(population[i]["genome"]);
    }
    return population;
}

function get_node_at_index(root, idx) {
    var unvisited_nodes = [root];
    var node = root;
    var cnt = 0;
    while (cnt <= idx && unvisited_nodes.length > 0) {
        node = unvisited_nodes.pop();
        if (typeof node !== 'string') {
            for (var i = node.length - 1; i > 0; i--) {
                unvisited_nodes.push(node[i]);
            }
        } else {
            unvisited_nodes.push(node);
        }
        cnt = cnt + 1;
    }
    console.assert(idx - cnt == -1, "idx not matching cnt - 1");
    return node;
}

function replace_subtree(new_subtree, old_subtree) {
    if (typeof old_subtree !== 'string') {
        old_subtree.splice(0, old_subtree.length);
        if (typeof new_subtree === 'string') {
            old_subtree.push(new_subtree);
        } else {
            for (var i = 0; i < new_subtree.length; i++) {
                old_subtree.push(new_subtree[i]);
            }
        }
    } else {
        old_subtree = new_subtree;
    }
}

function crossover(crossover_probability, population) {
    var CHILDREN = 2;
    var new_population = [];
    for (var i = 0; i < population.length; i = i + CHILDREN) {
        var children = [];
        for (var j = 0; j < CHILDREN; j++) {
            var idx = get_random_int(0, population.length);
            var genome = copy_tree(population[idx]["genome"]);
            var child = {
                genome: genome,
                fitness: DEFAULT_FITNESS
            };
            children.push(child);
        }
        if (get_random() < crossover_probability) {
            var xo_nodes = [];
            for (var j = 0; j < children.length; j++) {
                var end_node_idx = get_number_of_nodes(children[j]["genome"], 0) - 1;
                var node_idx = get_random_int(0, end_node_idx);
                xo_nodes.push(get_node_at_index(children[j]["genome"], node_idx));
            }
            var tmp_child_1_node = copy_tree(xo_nodes[1]);
            replace_subtree(xo_nodes[0], xo_nodes[1]);
            replace_subtree(tmp_child_1_node, xo_nodes[0]);
        }
        for (var j = 0; j < children.length; j++) {
            new_population.push(children[j]);
        }
    }
    return new_population;
}

function gp(params) {
    // Create population
    var population = initialize_population(params['population_size'],
        params['max_size']);
    evaluate_fitness(population);

    // Generation loop
    var generation = 0;
    while (generation < params['generations']) {
        // Selection
        var new_population = tournament_selection(params['tournament_size'], population);
        new_population = crossover(params['crossover_probability'], new_population);
        mutation(params['mutation_probability'], new_population, params['max_size']);

        // Evaluate the new population
        evaluate_fitness(new_population);

        // Replace the population with the new population
        population = new_population;

        print_stats(generation, new_population);

        // Increase the generation
        generation = generation + 1;
    }
}

var gp_params = {
    population_size: 8,
    max_size: 3,
    generations: 4,
    mutation_probability: 1.0,
    tournament_size: 2,
    crossover_probability: 1.0
};
// TODO fix size for mutation and crossover, it bloats too easily for mutation
gp(gp_params);

File: EC/python/pony_gp/pony_gp.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-


import cgi
import cgitb
import database
import random
import json
import tron_v3
import math

def tron_evaluate_AIs(AI_1, AI_2):
    tron_game = tron_v3.Tron(20, AI_1['genome'], AI_2['genome'])
    result = tron_game.run()
    #return 1 if AI2 wins, returns 0 if AI1 wins
    AI_1['fitness'] = AI_1['fitness'] - result + 0.1
    AI_2['fitness'] += float(result)/10
    

def get_symbols():
    arity = {
        "TURN_LEFT": 0,
        "TURN_RIGHT": 0,
        "SENSE_A": 0,
        "SENSE_L": 0,
        "SENSE_R": 0,
        "0.1": 0,
        "0.3": 0,
        "0.6": 0,
        "+": 2,
        "-": 2,
        "IFLEQ": 4
    };

    terminals = [];
    functions = [];

    for key in arity.keys():
        if (arity[key] == 0):
            terminals.append(key)
        else:
            functions.append(key)

    return {'arity': arity, 'terminals': terminals, 'functions': functions}

symbols = get_symbols();

DEFAULT_FITNESS = 0;

#returns a random symbol
#@param depth: current depth in tree
#@param max_depth: maximum depth
#@param symbols: total possible symbols
#@param full: whether the tree is full or not
def get_random_symbol(depth, max_depth, symbols, full):
    if (depth >= (max_depth - 1)):
        symbol = symbols["terminals"][random.randint(0, len(symbols["terminals"])-1)]
    else:
        terminal = get_random_boolean()
        if (not full and terminal):
            symbol = symbols["terminals"][random.randint(0, len(symbols["terminals"])-1)]
        else:
            symbol = symbols["functions"][random.randint(0, len(symbols["functions"])-1)]
    return symbol

def get_random_boolean():
    return random.random() < 0.5

#Adds a symbol to a node
#@param the tree to add the symbol to
#@param symbol: the symbol
#@param symbols: the collection of symbols, used to test if the symbol is terminal/functional.
def append_symbol(node, symbol, symbols):
    if symbol in symbols['terminals']:
        new_node = symbol;
    else:
        new_node = [symbol];
    node.append(new_node);
    return new_node
    
    
def append_symbol(node, symbol, symbols):
    if symbol in symbols['terminals']:
        new_node = symbol
    else:
        new_node = [symbol]
    node.append(new_node)
    return new_node

    
    
def grow(tree, depth, max_depth, full, symbols):
    if (type(tree) is not str):
        symbol = tree[0];
    else:
        symbol = tree;
    i = 0
    while (i < symbols["arity"][symbol]):
        new_symbol = get_random_symbol(depth, max_depth, symbols, full);
        new_node = append_symbol(tree, new_symbol, symbols);
        new_depth = depth + 1;
        if (new_symbol in symbols['functions']):
            grow(new_node, new_depth, max_depth, full, symbols)
        i += 1
    '''   
    def find_and_replace_subtree(root, subtree, node_idx, idx):
    idx = idx + 1;
    if (node_idx == idx):
        print 'root', root
        replace_subtree(subtree, root);
    else:
        for child in root[1:]:
            print 'child', child
            print type(child)
            idx = find_and_replace_subtree(child, subtree, node_idx, idx)
    return idx;
    '''
def find_and_replace_subtree(root, subtree, node_idx, idx):
    """
    Returns the current index and replaces the root with another subtree at the
    given index. The index is based on depth-first left-to-right traversal.

    :param root: Root of the tree
    :type root: list
    :param subtree: Subtree that will replace the root
    :type subtree: list
    :param node_idx: Index of the node to be replaced
    :type node_idx: int
    :param idx: Current index
    :type idx: int
    :return: Current index
    :rtype: int
    """

    # Increase the index
    print 'idx', idx
    idx += 1
    # Check if index is a the given node
    if node_idx == idx:
        print 'root', root
        if type(root) == str:
            print 'replacing', subtree
            root = subtree
            print subtree
        else:
            replace_subtree(subtree, root);
    elif type(root) == list:
        # Iterate over the children
        for child in get_children(root):
            print 'child', child
            # Recursively travers the child
            idx = find_and_replace_subtree(child, subtree, node_idx, idx)
    return idx
    


#convert a node_idx to a list of indexes that point to the index.
def find_subtree_index(subtree,node_idx, count, result):
    for sub_subtree in subtree:
        c

        
    
    
    
    
    
def get_children(node):
    """
    Return the children of the node. The children are all the elements of the
    except the first
    :param node: The node
    :type node: list
    :return: The children of the node
    :rtype: list
    """
    # Take a slice of the list except the head
    return node[1:]
    
    
        
def get_number_of_nodes(root, cnt):
    cnt = cnt + 1;
    if (type(root) is not str):
        i=1
        while (i < len(root)):
            cnt = get_number_of_nodes(root[i], cnt)
            i+=1
    return cnt

def get_depth_from_index(node, tree_info, node_idx, depth):
    if (node_idx == tree_info["idx"]):
        tree_info["idx_depth"] = depth
    tree_info["idx"] = tree_info["idx"] + 1
    i=1
    while i < len(node):
        depth = depth + 1;
        tree_info = get_depth_from_index(node[i], tree_info, node_idx, depth)
        depth = depth - 1
        i+=1
    return tree_info

def replace_subtree(new_subtree, old_subtree):
    print 'old', old_subtree, 'new', new_subtree
    if (type(old_subtree) is not str):
        del old_subtree[:]
        #http://stackoverflow.com/questions/1400608/how-to-empty-a-list-in-python
        if (type(new_subtree) is str):
            old_subtree.append(new_subtree)
        else:
            i=0
            while (i < len(new_subtree)):
                old_subtree.append(new_subtree[i])
                i+=1
    else:
        old_subtree = new_subtree;
        
def changeItem(listOfLists,element,changeTo):
  stringForm=repr(listOfLists)
  

class Tron_GA_v3(object):
    
    """
    TODO:
        READ DATABASE
        EVOLVE INDIVIDUALS FROM TRON GAME
        SAVE TO DATABASE
        NOT CLOG UP THREADS.
    
    
    
    """


    def __init__(self, population_size, max_size, generations, elite_size, crossover_probability, mutation_probability,
                 fitness_function):
        """Constructor

        

        :param population_size: Size of population
        :type population_size: integer
        :param max_size: Bitstring size for an individual solution
        :type max_size: integer
        :param generations: Number of iterations of the search loop
        :type generations: integer
        :param elite_size: Number of individuals preserved between generations
        :type elite_size: integer
        :param crossover_probability: Probability of crossing over two solutions
        :type crossover_probability: float
        :param mutation_probability: Probability of mutating a solution
        :type mutation_probability: float
        :param fitness_function: Method used to evaluate fitness of a solution
        :type fitness_function: Object
        """
        # Population size is the number of individual solutions
        self.population_size = population_size
        # Size of the individual solution
        self.max_size = max_size
        # Number of iterations of the search loop
        self.generations = generations
        # Number of individual solutions that are preserved between
        # generations
        self.elite_size = elite_size
        # Probability of crossover
        self.crossover_probability = crossover_probability
        # Probability of mutation
        self.mutation_probability = mutation_probability
        # Function that is used to evaluate the fitness of the
        # individual solution
        self.fitness_function = fitness_function

    def initialize_population(self):
        """//Initializes a Tron AI population
        //@param population_size size of total # of AI's
        //@param max_size Max Depth of AI's"""
        self.population = []
        i = 0
        while(i < self.population_size):
            full = random.randint(0,1)
            max_depth = (i % self.max_size) + 1;
            symbol = get_random_symbol(1, max_depth, symbols, full);
            tree = [symbol];
            if (max_depth > 0 and symbol in symbols["functions"]):
                grow(tree, 1, max_depth, full, symbols);
            self.population.append({'genome': tree, 'fitness': DEFAULT_FITNESS})
            print i
            print json.dumps(self.population[i]['genome'])
            i += 1
        return self.population;
        
        
        
    #TOOD: READ POPULATION FROM DB


    def search_loop(self, new_individuals):
        """
        Return the best individual from the evolutionary search
        loop. Starting from the initial population.

        

        :param new_individuals: Initial population of individuals
        :type new_individuals: List of Individual
        :returns: Best individual
        :rtype: Individual
        """

        #Initalise first generation
        individuals = []
        best_ever = None
        #Generation loop
        generation = 0
        while generation < self.generations:
            # Evaluate fitness
            self.evaluate_fitness(new_individuals, self.fitness_function)
            # Find best solution by sorting the population
            new_individuals.sort(reverse=True)
            best_ever = new_individuals[0]
            # Replace population
            individuals = self.generational_replacement(new_individuals,
                                                        individuals)
            # Print the stats of the population
            self.print_stats(generation, individuals)
            # Write population to file
            self.serialize_population(individuals)
            # Increase the generation counter
            generation += 1
            # Selection
            parents = self.tournament_selection(individuals)
            # Create new population
            new_individuals = []
            #TODO crossover does the copying of the individuals, is this too
            # implicit?
            while len(new_individuals) < self.population_size:
                # Vary the population by crossover
                new_individuals.extend(
                    # Pick 2 parents and pass them into crossover.
                    # '*' means that the list is collapsed
                    self.onepoint_crossover(*random.sample(parents, 2))
                )

            # Select population size individuals. Handles uneven population
            # sizes, since crossover returns 2 offspring
            new_individuals = new_individuals[:self.population_size]
            # Vary the population by mutation
            new_individuals = list(map(self.mutation, new_individuals))

        return best_ever

    def print_stats(self, generation, individuals):
        """
        Print the statistics for the generation and individuals.
       
        :param generation:generation number
        :type generation: integer
        :param individuals: population to get statistics for
        :type individuals: list of individuals
        """

        def get_ave_and_std(values):
            """
            Return average and standard deviation.            

            :param values: Values to calculate on
            :type values: list of numbers
            :returns: Average and Standard deviation of the input values
            :rtype: Tuple of floats
            """
            _ave = float(sum(values)) / len(values)
            _std = math.sqrt(float(
                sum((value - _ave) ** 2 for value in values)) / len(values))
            return _ave, _std

        # Sort individuals
        individuals.sort(reverse=True)
        # Get the fitness values
        fitness_values = [i.fitness for i in individuals]
        # Calculate average and standard deviation of the fitness in
        # the population
        ave_fit, std_fit = get_ave_and_std(fitness_values)
        # Print the statistics, including the best solution
        print("Gen:%d evals:%d fit_ave:%.2f+-%.3f %s" %
              (generation, (self.population_size * generation),
               ave_fit, std_fit,
               individuals[0]))


    def tournament_selection(self,tournament_size=2):
        """
        Return individuals from a population by drawing
        `tournament_size` competitors randomly and selecting the best
        of the competitors. `population_size` number of tournaments are
        held.

        :param population: Population to select from
        :type population: list of individuals
        :param tournament_size: Size of a tournament
        :type tournament_size: integer
        :returns: selected individuals
        :rtype: list of individuals
        """

        # Iterate until there are enough tournament winners selected
        winners = []
        while len(winners) < self.population_size:
            # Randomly select tournament size individual solutions
            # from the population.
            competitors = random.sample(self.population, tournament_size)
            # Rank the selected solutions
            competitors.sort(lambda x,y: x.fitness - y.fitness)
            # Append the best solution to the winners
            winners.append(competitors[0])
        return winners

    def generational_replacement(self, new_population, old_population):
        """
        Return new a population. The `elite_size` best old_population
        are appended to the new population. They are kept in the new
        population if they are better than the worst.

        :param new_population: the new population
        :type new_population: list of old_population
        :param old_population: the old population
        :type old_population: list of old_population
        :returns: the new population with the best from the old population
        :rtype: list of old_population
        """

        # Sort the population
        old_population.sort(reverse=True)
        # Append a copy of the best solutions of the old population to
        # the new population. ELITE_SIZE are taken
        for ind in old_population[:self.elite_size]:
            new_population.append(copy.deepcopy(ind))
        # Sort the new population
        new_population.sort(reverse=True)
        # Set the new population size
        return new_population[:self.population_size]

    @classmethod
    def evaluate_fitness(self,individuals, fitness_function):
        """
        Perform the fitness evaluation for each
        individual.

        :param individuals: Population to evaluate
        :type individuals: List of individuals
        :param fitness_function: Fitness function to evaluate the population on
        :type fitness_function: Object
        """
        # Iterate over all the individual solutions
        i =0 
        if len(individuals) % 2 == 1:
            individuals.append({'genome': ['TURN_LEFT'], 'fitness': DEFAULT_FITNESS})
        while i < len(individuals):
            fitness_function(individuals[i], individuals[i+1])
            print individuals[i]['fitness'], individuals[i+1]['fitness']
            i += 2

#WHAT IS GET_DEPTH_FROM_INDEX, WHAT IS FIND_AND_REPLACE_SUBTREE       
    def mutation(self,mutation_probability, individual):
        if (random.random() < mutation_probability):
            end_node_idx = get_number_of_nodes(individual["genome"], 0) - 1;
            print 'end index,' 
            print end_node_idx
            node_idx =random.randint(0, end_node_idx);
            node_info = get_depth_from_index(individual["genome"], {'idx_depth': 0, 'idx': 0}, node_idx, 0);
            max_subtree_depth = self.max_size - node_info['idx_depth'];
            new_subtree = [get_random_symbol(max_subtree_depth, self.max_size, symbols,False)]
            print new_subtree
            print node_idx
            if (new_subtree[0] in symbols['functions']):
                full = False
                grow(new_subtree, node_info['idx_depth'], self.max_size, full, symbols)
            find_and_replace_subtree(individual["genome"], new_subtree, node_idx, -1)


    def onepoint_crossover(self, parent_0, parent_1):
        """
        Return a list of two new individuals that given two
        individuals, create two children using one-point crossover.

        :param parent_0: Parent one
        :type parent_0: Individual
        :param parent_1: Parent two
        :type parent_1: Individual
        :returns: Two children from the parents
        :rtype: List of individuals
        """
        # Get the genomes from the parents
        genome_parent_0, genome_parent_1 = parent_0.genome, parent_1.genome
        # Uniformly generate crossover points. 
        pt_p = random.randint(1, self.max_size)
        # Make new chromosomes by crossover.
        if random.random() < self.crossover_probability:
            child_0 = genome_parent_0[:pt_p] + genome_parent_1[pt_p:]
            child_1 = genome_parent_1[:pt_p] + genome_parent_0[pt_p:]
        else:
            child_0, child_1 = genome_parent_0[:], genome_parent_1[:]
        # Put the new chromosomes into new individuals
        return [Individual(child_0), Individual(child_1)]

    def run(self):
        """
        Return the best solution. Create an initial
        population. Perform an evolutionary search.

        :returns: Best solution
        :rtype: Individual
        """

        #Create population
        individuals = self.initialize_population()
        # Start evolutionary search
        best_ever = self.search_loop(individuals)

        return best_ever







print "Content-Type: text/plain;charset=utf-8"
print 
print "Hello World!"


t=Tron_GA_v3(10,4,10,10,0,0,1)
t.run()
t.initialize_population()
t.evaluate_fitness(t.population, tron_evaluate_AIs)
print t.population[2]
t.mutation(1,t.population[2])
print t.population[2]
